### ISSUE TITLE
Provide a way to disable the "Consider using a tolerance" warning when matching exact doubles

### ISSUE THREAD

**Issue by Hikaru755** [kotlintest/kotlintest#59]
Whenever you compare two values like this: `d shouldBe 0.0` you get this warning:

> [WARN] When comparing doubles consider using tolerance, eg: a shouldBe b plusOrMinus c

I have a project with a lot of testcases that do indeed need exact double matches with no tolerance, and this output spamming the console for every double comparison is kind of annoying. I would appreciate if there was a way to disable that output - a possible way that goes in line with the testing syntax might be to use a syntax like this: `d shouldBe exactly(0.0)`

**Comment by sksamuel** [comment]
:+1:

**Comment by sksamuel** [comment]
I've pushed this out as version 1.3.2

`a shouldBe exactly(b)`

**Comment by helmbold** [comment]
Following semantic versioning this should have gone to 1.4 ;-)

**Comment by sksamuel** [comment]
Ah yes suppose so.

**Comment by russel** [comment]
Are you sure you really mean you want exact matching of two hardware float values? BigDecimal values clearly I can appreciate wanting exact matching, but for hardware float and double? I am genuinely surprised anyone would think this useful.

**Comment by Hikaru755** [comment]
@russel In my use case I simply want to check that some double values are being passed around in a model correctly, without any math happening in between. Having to specify an arbitrary tolerance everytime for those tests just makes them harder to read without adding benefits, and if I don't, well, I get these error messages spamming the console unnecessarily.

**Comment by russel** [comment]
@Hikaru755 I have admit I find the idea of wanting to test that a double value is passed around correctly either a failure of the application architecture or a misuse of tests. I don't know your codebase, but I worry you are getting the wrong solution to the wrong problem with this "fix" to the matchers.

@sksamuel I feel have an "exactly" matcher is going to lead a lot of people into a lot of problems. The idea of one hardware floating point number being exactly the same as another when the values are expressed in base 10 in source and represented as base 2 in the executable is fundamentally problematic. Having this matcher is going to give a lot of people the wrong impression of floating point and matching. Having a default match range for floating point numbers and assuming the programmers are not being stupid is a far better solution to the problem of messages on the screen. OK so this will mask any faults in @Hikaru755 use case, but I have problems with that use of tests anyway so I am not that worried.

As you can tell I believe this changeset should be revoked.

**Comment by Hikaru755** [comment]
@russel So assuming I'm doing this in one of my tests:
```
foo = 1.0
bar()
foo shouldBe 1.0
```
What would you do differently about that? Basically I just want to make sure that the call to bar() didn't affect the foo property. I don't really see yet how that would be failure of the architecture or test misuse. Or am I just missing something and can't reasonably assume that the 1.0 I put into foo won't be the same as the 1.0 I'm comparing it to?

**Comment by sksamuel** [comment]
`println(1.03 - 0.42 == 0.61)` prints false as an example of dodgy doubles.

But for straightforward assignment you are ok. So if you're literally just checking that x hasn't changed, you can do `==`.

For exact equality, you can also use 
`Double.doubleToLongBits(x) == Double.doubleToLongBits(y)`

**Comment by sksamuel** [comment]
@russel Your idea of having a matcher that has a default tolerance is a good idea. I'm not in favour of hiding it through, I'd want something like `x shouldBe roughly(y)` but perhaps with a better word than roughly, one that people are happy to use by default. And the exactly(y) case can be used purely for bitwise equality.

**Comment by helmbold** [comment]
I'd vote against `exactly` since it is generally not useful and perhaps misleading. And there is ==. Maybe we can have something like [ScalaTest]( http://www.scalatest.org/user_guide/using_matchers#checkingNumbersAgainstARange) has. Or maybe `between `?

**Comment by sksamuel** [comment]
`a shouldBe between(0,1)` is quite a nice syntax anyway, but not that great for `a shouldBe between(-0.01,0.01)`

**Comment by russel** [comment]
@Hikaru755 Yes I have serious doubts about this example:
```
foo = 1.0
bar()
foo shouldBe 1.0
```
to me it shows bad testing, or worse testing for the sake of testing. Functions/procedures/subroutines whatever name, are units of abstraction, and units of scope. By making a call, `bar()`, you are entering a new scope. The symbol `foo` should be single assignment, but let's assume it is a variable. It is not in the scope of the called function. (There appear to be no closures here so we are in a very simple situation.) Thus, if the variable `foo` can be changed by the execution of `bar`, either the program is very, very, very badly designed and should be scrapped, or the programming language is seriously broken, in which case the code should be scrapped and rewritten in a better language. Thus in all cases this test is not achieving anything useful. What aspect of the external activity is being tested here? What is the problem that this test solves that is outward facing? This all leads to this test being fundamentally redundant.

So given the use case that justified the exact match is not a valid use case, I say there is no justification for the exact match and it should be removed because it's presence is dangerous to the way people test floating point numbers.

**Comment by russel** [comment]
@sksamuel I quite like roughly on a British humour basis, but it is vague – which is the point! the between idea works for me in an abstract way, and for some use cases, but will get difficult when you want 14 significant digits of accuracy. Maybe take a leaf out of the Fortran (not FORTRAN) book and have a way of saying "the same to N significant digits". This directly expresses what is actually being tested. Python almost got this right with their almostEqual method, buit it has a fundamental problem that I am surprised the Python hierarchy let through. So I would suggest the core test be equalToNDigits(n) with a default of say 9 for float and 12 for double.

**Comment by Hikaru755** [comment]
@russel Regarding this explicit example I'd agree with you, perhaps I oversimplified it for the sake of readability. So let's make the example a little bit more elaborate and closer to what I'm actually doing:
```
someObject.foo = 1.0
someObject.bar()
someObject.foo shouldBe 1.0
```
In this case foo is a property of someObject, and therefore in scope of the method bar of the same object. Imagine that for example the call to bar() should not change the property in this case, but there could be another context for this where the call should actually be doing something to the variable (some flag set on the object, whatever). So here I want to make sure that in **this** case, that doesn't happen. Yes, I know it's not the clean functional approach to have a method work with possibly obscure side effects like these, but that's the reality of imperative and object oriented languages and frameworks, especially when dealing with legacy codebases.

I mean, would you argue that no one should produce hammers anymore, just because there will always be users who don't know how to use them and hit themselves with them? I agree that cases like this should not be all that common or strived for, but when you have them, I think it's better to have the option than not. And to be honest, I don't think a library like this is there to educate its users about good practises, at least not at the expense of keeping potentially useful functions even from those that know how to use them. It's a good idea to mention this problem for example in the documentation for the exactly() method though.

**Comment by sksamuel** [comment]
@Hikaru755 I think `exactly` is what you want in this case. I agree with your testing approach.

@all Testing that a double is equal to some specific value is fine. It's only problematic when you are testing the result of some equation. Then the result might not be the same as the value you expect because of representation roundoff.

So, this is fine:

```
val x  = 1.0
x shouldBe exactly(1.0)

val y = 1.0
val z = 1.0 + 1.5
z shouldBe (2.5 plusOrMinus 0.01)

val a = 1.0
val b = 1.0 + 1.5
b shouldBe someKeywordLikeRoughlyOrAlmost(2.5)
```

This is not:

```
val x = 1.0
val y = 1.0 + 1.5
y shouldBe 2.5
```

I think we're all in agreement here. We all know how doubles work :) We just need to perhaps add a new keyword like `roughly` as a shortcut for `shouldBe (x plusOrMinus someSmallValue)` for people who are genuinely testing the results of floating point arithmatic (which @Hikaru755 is not). Ideas on the keyword appreciated.

**Comment by Hikaru755** [comment]
Please put `someKeywordLikeRoughlyOrAlmost(x)` in there as a synonym to whatever keyword will be final.

I like `almostEqual` or `roughly` for the actual keyword though. And @russel's suggestion of equalToNDigits is another good suggestion for other tests as a more intuitive version of plusOrMinus - maybe in this syntax:
```kotlin
a shouldEqual 1.0 to 12.digits
```
I don't know exactly how the matchers work internally as I haven't looked too closely there yet, but from the existing syntax something in this vein could work and is quite easy to read.

**Comment by russel** [comment]
I am fairly sure Cope would say that any testing of internal object behaviours is a waste of time, not worth it. There are many threads and videos on the internet saying effectively that only end-to-end tests are worth anything at all. I have to admit I am a little bit in this camp: too many programmers get appropriately obsessed, but all too often about the wrong tests.

Having said that, now that the example has shifted to testing methods of an object, and so the scope thing has gone out of the window, there is in principle a case to answer: the execution of method `bar` should not affect the state of `foo`. I still have this niggle that exactly is not the correct tool, that perhaps some notion of unchanged would be better. I guess my worry here is that you have tested the same value, not that foo is unaltered. I wonder if the only way of checking this is some form of mock/reflection test.
